{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyO6Wxt7t9ApoFm0uR6xl2R4",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/escobedolozoya/vhlab_vhtools/blob/master/RNAseq_Analysis.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "vqW__brQ7TIa",
        "outputId": "26f3189d-b239-4cf5-b9b1-367fa61f4f87"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Extracting templates from packages: 100%\n",
            "Selecting previously unselected package default-jre-headless.\n",
            "(Reading database ... 126455 files and directories currently installed.)\n",
            "Preparing to unpack .../00-default-jre-headless_2%3a1.11-72build2_amd64.deb ...\n",
            "Unpacking default-jre-headless (2:1.11-72build2) ...\n",
            "Selecting previously unselected package libxtst6:amd64.\n",
            "Preparing to unpack .../01-libxtst6_2%3a1.2.3-1build4_amd64.deb ...\n",
            "Unpacking libxtst6:amd64 (2:1.2.3-1build4) ...\n",
            "Selecting previously unselected package openjdk-11-jre:amd64.\n",
            "Preparing to unpack .../02-openjdk-11-jre_11.0.28+6-1ubuntu1~22.04.1_amd64.deb ...\n",
            "Unpacking openjdk-11-jre:amd64 (11.0.28+6-1ubuntu1~22.04.1) ...\n",
            "Selecting previously unselected package default-jre.\n",
            "Preparing to unpack .../03-default-jre_2%3a1.11-72build2_amd64.deb ...\n",
            "Unpacking default-jre (2:1.11-72build2) ...\n",
            "Selecting previously unselected package libcommons-math3-java.\n",
            "Preparing to unpack .../04-libcommons-math3-java_3.6.1-3_all.deb ...\n",
            "Unpacking libcommons-math3-java (3.6.1-3) ...\n",
            "Selecting previously unselected package libjbzip2-java.\n",
            "Preparing to unpack .../05-libjbzip2-java_0.9.1-7_all.deb ...\n",
            "Unpacking libjbzip2-java (0.9.1-7) ...\n",
            "Selecting previously unselected package libapache-pom-java.\n",
            "Preparing to unpack .../06-libapache-pom-java_18-1_all.deb ...\n",
            "Unpacking libapache-pom-java (18-1) ...\n",
            "Selecting previously unselected package libcommons-parent-java.\n",
            "Preparing to unpack .../07-libcommons-parent-java_43-1_all.deb ...\n",
            "Unpacking libcommons-parent-java (43-1) ...\n",
            "Selecting previously unselected package libcommons-compress-java.\n",
            "Preparing to unpack .../08-libcommons-compress-java_1.21-1_all.deb ...\n",
            "Unpacking libcommons-compress-java (1.21-1) ...\n",
            "Selecting previously unselected package libcommons-logging-java.\n",
            "Preparing to unpack .../09-libcommons-logging-java_1.2-2_all.deb ...\n",
            "Unpacking libcommons-logging-java (1.2-2) ...\n",
            "Selecting previously unselected package libcommons-jexl2-java.\n",
            "Preparing to unpack .../10-libcommons-jexl2-java_2.1.1-5_all.deb ...\n",
            "Unpacking libcommons-jexl2-java (2.1.1-5) ...\n",
            "Selecting previously unselected package libjson-simple-java.\n",
            "Preparing to unpack .../11-libjson-simple-java_2.3.0-1_all.deb ...\n",
            "Unpacking libjson-simple-java (2.3.0-1) ...\n",
            "Selecting previously unselected package libngs-sdk2:amd64.\n",
            "Preparing to unpack .../12-libngs-sdk2_2.11.2-3_amd64.deb ...\n",
            "Unpacking libngs-sdk2:amd64 (2.11.2-3) ...\n",
            "Selecting previously unselected package libngs-sdk-dev:amd64.\n",
            "Preparing to unpack .../13-libngs-sdk-dev_2.11.2-3_amd64.deb ...\n",
            "Unpacking libngs-sdk-dev:amd64 (2.11.2-3) ...\n",
            "Selecting previously unselected package libngs-java:amd64.\n",
            "Preparing to unpack .../14-libngs-java_2.11.2-3_amd64.deb ...\n",
            "Unpacking libngs-java:amd64 (2.11.2-3) ...\n",
            "Selecting previously unselected package libsnappy-jni.\n",
            "Preparing to unpack .../15-libsnappy-jni_1.1.8.3-1_amd64.deb ...\n",
            "Unpacking libsnappy-jni (1.1.8.3-1) ...\n",
            "Selecting previously unselected package libsnappy-java.\n",
            "Preparing to unpack .../16-libsnappy-java_1.1.8.3-1_all.deb ...\n",
            "Unpacking libsnappy-java (1.1.8.3-1) ...\n",
            "Selecting previously unselected package libhtsjdk-java.\n",
            "Preparing to unpack .../17-libhtsjdk-java_2.24.1+dfsg-1_all.deb ...\n",
            "Unpacking libhtsjdk-java (2.24.1+dfsg-1) ...\n",
            "Selecting previously unselected package libfindbin-libs-perl.\n",
            "Preparing to unpack .../18-libfindbin-libs-perl_2.190.02-1_all.deb ...\n",
            "Unpacking libfindbin-libs-perl (2.190.02-1) ...\n",
            "Selecting previously unselected package libargs4j-java.\n",
            "Preparing to unpack .../19-libargs4j-java_2.33-1.1_all.deb ...\n",
            "Unpacking libargs4j-java (2.33-1.1) ...\n",
            "Selecting previously unselected package libcommons-io-java.\n",
            "Preparing to unpack .../20-libcommons-io-java_2.11.0-2_all.deb ...\n",
            "Unpacking libcommons-io-java (2.11.0-2) ...\n",
            "Selecting previously unselected package libcommons-lang3-java.\n",
            "Preparing to unpack .../21-libcommons-lang3-java_3.11-1_all.deb ...\n",
            "Unpacking libcommons-lang3-java (3.11-1) ...\n",
            "Selecting previously unselected package libsis-base-jni.\n",
            "Preparing to unpack .../22-libsis-base-jni_18.09~pre1+git20180928.45fbd31+dfsg-3_amd64.deb ...\n",
            "Unpacking libsis-base-jni (18.09~pre1+git20180928.45fbd31+dfsg-3) ...\n",
            "Selecting previously unselected package libsis-base-java.\n",
            "Preparing to unpack .../23-libsis-base-java_18.09~pre1+git20180928.45fbd31+dfsg-3_all.deb ...\n",
            "Unpacking libsis-base-java (18.09~pre1+git20180928.45fbd31+dfsg-3) ...\n",
            "Selecting previously unselected package libsis-jhdf5-jni.\n",
            "Preparing to unpack .../24-libsis-jhdf5-jni_19.04.0+dfsg-4_amd64.deb ...\n",
            "Unpacking libsis-jhdf5-jni (19.04.0+dfsg-4) ...\n",
            "Selecting previously unselected package libsis-jhdf5-java.\n",
            "Preparing to unpack .../25-libsis-jhdf5-java_19.04.0+dfsg-4_all.deb ...\n",
            "Unpacking libsis-jhdf5-java (19.04.0+dfsg-4) ...\n",
            "Selecting previously unselected package fastqc.\n",
            "Preparing to unpack .../26-fastqc_0.11.9+dfsg-5_all.deb ...\n",
            "Unpacking fastqc (0.11.9+dfsg-5) ...\n",
            "Selecting previously unselected package fonts-dejavu-core.\n",
            "Preparing to unpack .../27-fonts-dejavu-core_2.37-2build1_all.deb ...\n",
            "Unpacking fonts-dejavu-core (2.37-2build1) ...\n",
            "Selecting previously unselected package fonts-dejavu-extra.\n",
            "Preparing to unpack .../28-fonts-dejavu-extra_2.37-2build1_all.deb ...\n",
            "Unpacking fonts-dejavu-extra (2.37-2build1) ...\n",
            "Selecting previously unselected package libxxf86dga1:amd64.\n",
            "Preparing to unpack .../29-libxxf86dga1_2%3a1.1.5-0ubuntu3_amd64.deb ...\n",
            "Unpacking libxxf86dga1:amd64 (2:1.1.5-0ubuntu3) ...\n",
            "Selecting previously unselected package x11-utils.\n",
            "Preparing to unpack .../30-x11-utils_7.7+5build2_amd64.deb ...\n",
            "Unpacking x11-utils (7.7+5build2) ...\n",
            "Selecting previously unselected package libatk-wrapper-java.\n",
            "Preparing to unpack .../31-libatk-wrapper-java_0.38.0-5build1_all.deb ...\n",
            "Unpacking libatk-wrapper-java (0.38.0-5build1) ...\n",
            "Selecting previously unselected package libatk-wrapper-java-jni:amd64.\n",
            "Preparing to unpack .../32-libatk-wrapper-java-jni_0.38.0-5build1_amd64.deb ...\n",
            "Unpacking libatk-wrapper-java-jni:amd64 (0.38.0-5build1) ...\n",
            "Setting up default-jre-headless (2:1.11-72build2) ...\n",
            "Setting up libxtst6:amd64 (2:1.2.3-1build4) ...\n",
            "Setting up libngs-sdk2:amd64 (2.11.2-3) ...\n",
            "Setting up libxxf86dga1:amd64 (2:1.1.5-0ubuntu3) ...\n",
            "Setting up openjdk-11-jre:amd64 (11.0.28+6-1ubuntu1~22.04.1) ...\n",
            "Setting up libfindbin-libs-perl (2.190.02-1) ...\n",
            "Setting up default-jre (2:1.11-72build2) ...\n",
            "Setting up libjson-simple-java (2.3.0-1) ...\n",
            "Setting up libsis-base-jni (18.09~pre1+git20180928.45fbd31+dfsg-3) ...\n",
            "Setting up libapache-pom-java (18-1) ...\n",
            "Setting up fonts-dejavu-core (2.37-2build1) ...\n",
            "Setting up libjbzip2-java (0.9.1-7) ...\n",
            "Setting up libngs-sdk-dev:amd64 (2.11.2-3) ...\n",
            "Setting up fonts-dejavu-extra (2.37-2build1) ...\n",
            "Setting up x11-utils (7.7+5build2) ...\n",
            "Setting up libcommons-math3-java (3.6.1-3) ...\n",
            "Setting up libsis-jhdf5-jni (19.04.0+dfsg-4) ...\n",
            "Setting up libatk-wrapper-java (0.38.0-5build1) ...\n",
            "Setting up libngs-java:amd64 (2.11.2-3) ...\n",
            "Setting up libsnappy-jni (1.1.8.3-1) ...\n",
            "Setting up libargs4j-java (2.33-1.1) ...\n",
            "Setting up libcommons-parent-java (43-1) ...\n",
            "Setting up libcommons-logging-java (1.2-2) ...\n",
            "Setting up libatk-wrapper-java-jni:amd64 (0.38.0-5build1) ...\n",
            "Setting up libcommons-jexl2-java (2.1.1-5) ...\n",
            "Setting up libcommons-lang3-java (3.11-1) ...\n",
            "Setting up libsnappy-java (1.1.8.3-1) ...\n",
            "Setting up libcommons-compress-java (1.21-1) ...\n",
            "Setting up libcommons-io-java (2.11.0-2) ...\n",
            "Setting up libsis-base-java (18.09~pre1+git20180928.45fbd31+dfsg-3) ...\n",
            "Setting up libhtsjdk-java (2.24.1+dfsg-1) ...\n",
            "Setting up libsis-jhdf5-java (19.04.0+dfsg-4) ...\n",
            "Setting up fastqc (0.11.9+dfsg-5) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for mailcap (3.70+nmu1ubuntu1) ...\n",
            "Processing triggers for fontconfig (2.13.1-4.2ubuntu5) ...\n",
            "Processing triggers for hicolor-icon-theme (0.17-2) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero_v2.so.0 is not a symbolic link\n",
            "\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m46.3/46.3 kB\u001b[0m \u001b[31m1.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.7/8.7 MB\u001b[0m \u001b[31m57.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m79.9/79.9 MB\u001b[0m \u001b[31m11.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m139.3/139.3 kB\u001b[0m \u001b[31m10.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m46.0/46.0 kB\u001b[0m \u001b[31m3.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m39.6/39.6 MB\u001b[0m \u001b[31m19.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m70.2/70.2 kB\u001b[0m \u001b[31m5.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m14.1/14.1 MB\u001b[0m \u001b[31m97.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m40.7/40.7 kB\u001b[0m \u001b[31m2.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m86.8/86.8 kB\u001b[0m \u001b[31m6.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m85.7/85.7 kB\u001b[0m \u001b[31m6.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hPreconfiguring packages ...\n",
            "Selecting previously unselected package menu.\n",
            "(Reading database ... 127044 files and directories currently installed.)\n",
            "Preparing to unpack .../0-menu_2.1.47ubuntu4_amd64.deb ...\n",
            "Unpacking menu (2.1.47ubuntu4) ...\n",
            "Selecting previously unselected package blends-common.\n",
            "Preparing to unpack .../1-blends-common_0.7.4ubuntu1_all.deb ...\n",
            "Unpacking blends-common (0.7.4ubuntu1) ...\n",
            "Selecting previously unselected package libkdf5-2.\n",
            "Preparing to unpack .../2-libkdf5-2_2.11.2+dfsg-4build2_amd64.deb ...\n",
            "Unpacking libkdf5-2 (2.11.2+dfsg-4build2) ...\n",
            "Selecting previously unselected package libncbi-vdb2.\n",
            "Preparing to unpack .../3-libncbi-vdb2_2.11.2+dfsg-4build2_amd64.deb ...\n",
            "Unpacking libncbi-vdb2 (2.11.2+dfsg-4build2) ...\n",
            "Selecting previously unselected package libncbi-wvdb2.\n",
            "Preparing to unpack .../4-libncbi-wvdb2_2.11.2+dfsg-4build2_amd64.deb ...\n",
            "Unpacking libncbi-wvdb2 (2.11.2+dfsg-4build2) ...\n",
            "Selecting previously unselected package sra-toolkit.\n",
            "Preparing to unpack .../5-sra-toolkit_2.11.3+dfsg-1ubuntu1_amd64.deb ...\n",
            "Unpacking sra-toolkit (2.11.3+dfsg-1ubuntu1) ...\n",
            "Selecting previously unselected package med-config.\n",
            "Preparing to unpack .../6-med-config_3.7.1_all.deb ...\n",
            "Unpacking med-config (3.7.1) ...\n",
            "Setting up libncbi-vdb2 (2.11.2+dfsg-4build2) ...\n",
            "Setting up libkdf5-2 (2.11.2+dfsg-4build2) ...\n",
            "Setting up libncbi-wvdb2 (2.11.2+dfsg-4build2) ...\n",
            "Setting up menu (2.1.47ubuntu4) ...\n",
            "Setting up blends-common (0.7.4ubuntu1) ...\n",
            "Setting up sra-toolkit (2.11.3+dfsg-1ubuntu1) ...\n",
            "Setting up med-config (3.7.1) ...\n",
            "Adding group `med' (GID 107) ...\n",
            "Done.\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero_v2.so.0 is not a symbolic link\n",
            "\n",
            "Processing triggers for menu (2.1.47ubuntu4) ...\n",
            "FastQC v0.11.9\n",
            "multiqc, version 1.32\n"
          ]
        }
      ],
      "source": [
        "# Step 1: Install the tools we need\n",
        "!apt-get install -qq fastqc\n",
        "!pip install -q multiqc\n",
        "!apt-get install -qq sra-toolkit\n",
        "\n",
        "# Verify installations\n",
        "!fastqc --version\n",
        "!multiqc --version"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Create a working directory\n",
        "!mkdir -p fastq_files\n",
        "!mkdir -p fastqc_output\n",
        "\n",
        "# Download FIRST 100,000 reads from one sample (Control replicate 1)\n",
        "# This keeps the file small (~50MB instead of several GB)\n",
        "!fastq-dump --split-files -X 100000 --outdir fastq_files SAMN31104764\n",
        "\n",
        "# Check what we got\n",
        "!ls -lh fastq_files/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lNGJu87M9tQU",
        "outputId": "109f773c-c977-428a-a717-c1fc781b2932"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Rejected 100000 READS because READLEN < 1\n",
            "Read 100000 spots for SAMN31104764\n",
            "Written 100000 spots for SAMN31104764\n",
            "total 182M\n",
            "-rw-r--r-- 1 root root 22M Oct 30 21:11 SAMN31104764_1.fastq\n",
            "-rw-r--r-- 1 root root 22M Oct 30 20:57 SAMN31104768_1.fastq\n",
            "-rw-r--r-- 1 root root 36M Oct 30 16:33 SRR21820609_1.fastq\n",
            "-rw-r--r-- 1 root root 34M Oct 30 17:27 SRR21820609_1_trimmed.fastq\n",
            "-rw-r--r-- 1 root root 36M Oct 30 16:33 SRR21820609_2.fastq\n",
            "-rw-r--r-- 1 root root 34M Oct 30 17:27 SRR21820609_2_trimmed.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Run FastQC on both paired-end files\n",
        "!fastqc fastq_files/SRR21820609_1.fastq fastq_files/SRR21820609_2.fastq -o fastqc_output/\n",
        "\n",
        "# List what was generated\n",
        "!ls -lh fastqc_output/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pqCjvMR4AFQC",
        "outputId": "802693be-e1a2-4fe2-81a4-e26ba34d5174"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Started analysis of SRR21820609_1.fastq\n",
            "Approx 5% complete for SRR21820609_1.fastq\n",
            "Approx 10% complete for SRR21820609_1.fastq\n",
            "Approx 15% complete for SRR21820609_1.fastq\n",
            "Approx 20% complete for SRR21820609_1.fastq\n",
            "Approx 25% complete for SRR21820609_1.fastq\n",
            "Approx 30% complete for SRR21820609_1.fastq\n",
            "Approx 35% complete for SRR21820609_1.fastq\n",
            "Approx 40% complete for SRR21820609_1.fastq\n",
            "Approx 45% complete for SRR21820609_1.fastq\n",
            "Approx 50% complete for SRR21820609_1.fastq\n",
            "Approx 55% complete for SRR21820609_1.fastq\n",
            "Approx 60% complete for SRR21820609_1.fastq\n",
            "Approx 65% complete for SRR21820609_1.fastq\n",
            "Approx 70% complete for SRR21820609_1.fastq\n",
            "Approx 75% complete for SRR21820609_1.fastq\n",
            "Approx 80% complete for SRR21820609_1.fastq\n",
            "Approx 85% complete for SRR21820609_1.fastq\n",
            "Approx 90% complete for SRR21820609_1.fastq\n",
            "Approx 95% complete for SRR21820609_1.fastq\n",
            "Approx 100% complete for SRR21820609_1.fastq\n",
            "Analysis complete for SRR21820609_1.fastq\n",
            "Started analysis of SRR21820609_2.fastq\n",
            "Approx 5% complete for SRR21820609_2.fastq\n",
            "Approx 10% complete for SRR21820609_2.fastq\n",
            "Approx 15% complete for SRR21820609_2.fastq\n",
            "Approx 20% complete for SRR21820609_2.fastq\n",
            "Approx 25% complete for SRR21820609_2.fastq\n",
            "Approx 30% complete for SRR21820609_2.fastq\n",
            "Approx 35% complete for SRR21820609_2.fastq\n",
            "Approx 40% complete for SRR21820609_2.fastq\n",
            "Approx 45% complete for SRR21820609_2.fastq\n",
            "Approx 50% complete for SRR21820609_2.fastq\n",
            "Approx 55% complete for SRR21820609_2.fastq\n",
            "Approx 60% complete for SRR21820609_2.fastq\n",
            "Approx 65% complete for SRR21820609_2.fastq\n",
            "Approx 70% complete for SRR21820609_2.fastq\n",
            "Approx 75% complete for SRR21820609_2.fastq\n",
            "Approx 80% complete for SRR21820609_2.fastq\n",
            "Approx 85% complete for SRR21820609_2.fastq\n",
            "Approx 90% complete for SRR21820609_2.fastq\n",
            "Approx 95% complete for SRR21820609_2.fastq\n",
            "Approx 100% complete for SRR21820609_2.fastq\n",
            "Analysis complete for SRR21820609_2.fastq\n",
            "total 2.1M\n",
            "-rw-r--r-- 1 root root 634K Oct 30 16:43 SRR21820609_1_fastqc.html\n",
            "-rw-r--r-- 1 root root 468K Oct 30 16:43 SRR21820609_1_fastqc.zip\n",
            "-rw-r--r-- 1 root root 603K Oct 30 16:43 SRR21820609_2_fastqc.html\n",
            "-rw-r--r-- 1 root root 416K Oct 30 16:43 SRR21820609_2_fastqc.zip\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Download the HTML reports so we can open them\n",
        "from google.colab import files\n",
        "\n",
        "files.download('fastqc_output/SRR21820609_1_fastqc.html')\n",
        "files.download('fastqc_output/SRR21820609_2_fastqc.html')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "_SIrZFqTBhNR",
        "outputId": "4ff9aa98-0c4a-4f69-d090-483a8dd6c697"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_fd5dcbf1-609b-4a7d-b8a8-72f8aea624e2\", \"SRR21820609_1_fastqc.html\", 648218)"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_897adf45-6b92-43a0-aaa1-f6854adc9d3a\", \"SRR21820609_2_fastqc.html\", 616972)"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Install fastp\n",
        "!apt-get install -qq fastp\n",
        "\n",
        "# Run fastp on our test sample\n",
        "!fastp \\\n",
        "  -i fastq_files/SRR21820609_1.fastq \\\n",
        "  -I fastq_files/SRR21820609_2.fastq \\\n",
        "  -o fastq_files/SRR21820609_1_trimmed.fastq \\\n",
        "  -O fastq_files/SRR21820609_2_trimmed.fastq \\\n",
        "  -h fastqc_output/fastp_report.html \\\n",
        "  -j fastqc_output/fastp_report.json\n",
        "\n",
        "# Check output\n",
        "!ls -lh fastq_files/*trimmed*"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zABw4ecPKSBG",
        "outputId": "7a2100a7-405b-4a4c-9a60-3a1060800b71"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Selecting previously unselected package fastp.\n",
            "(Reading database ... 127394 files and directories currently installed.)\n",
            "Preparing to unpack .../fastp_0.20.1+dfsg-1_amd64.deb ...\n",
            "Unpacking fastp (0.20.1+dfsg-1) ...\n",
            "Setting up fastp (0.20.1+dfsg-1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Read1 before filtering:\n",
            "total reads: 100000\n",
            "total bases: 15100000\n",
            "Q20 bases: 14752216(97.6968%)\n",
            "Q30 bases: 14220908(94.1782%)\n",
            "\n",
            "Read2 before filtering:\n",
            "total reads: 100000\n",
            "total bases: 15100000\n",
            "Q20 bases: 14177887(93.8933%)\n",
            "Q30 bases: 13198541(87.4076%)\n",
            "\n",
            "Read1 after filtering:\n",
            "total reads: 95919\n",
            "total bases: 13966846\n",
            "Q20 bases: 13708025(98.1469%)\n",
            "Q30 bases: 13275301(95.0487%)\n",
            "\n",
            "Read2 aftering filtering:\n",
            "total reads: 95919\n",
            "total bases: 13966846\n",
            "Q20 bases: 13407992(95.9987%)\n",
            "Q30 bases: 12607341(90.2662%)\n",
            "\n",
            "Filtering result:\n",
            "reads passed filter: 191838\n",
            "reads failed due to low quality: 8100\n",
            "reads failed due to too many N: 62\n",
            "reads failed due to too short: 0\n",
            "reads with adapter trimmed: 38608\n",
            "bases trimmed due to adapters: 1034738\n",
            "\n",
            "Duplication rate: 3.80623%\n",
            "\n",
            "Insert size peak (evaluated by paired-end reads): 158\n",
            "\n",
            "JSON report: fastqc_output/fastp_report.json\n",
            "HTML report: fastqc_output/fastp_report.html\n",
            "\n",
            "fastp -i fastq_files/SRR21820609_1.fastq -I fastq_files/SRR21820609_2.fastq -o fastq_files/SRR21820609_1_trimmed.fastq -O fastq_files/SRR21820609_2_trimmed.fastq -h fastqc_output/fastp_report.html -j fastqc_output/fastp_report.json \n",
            "fastp v0.20.1, time used: 3 seconds\n",
            "-rw-r--r-- 1 root root 34M Oct 30 17:27 fastq_files/SRR21820609_1_trimmed.fastq\n",
            "-rw-r--r-- 1 root root 34M Oct 30 17:27 fastq_files/SRR21820609_2_trimmed.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Filter the GTF to only chr19 (using correct filename)\n",
        "!grep \"^chr19\" reference/mm39_genes.gtf > reference/chr19_genes.gtf\n",
        "\n",
        "# Verify it worked\n",
        "!wc -l reference/chr19_genes.gtf\n",
        "!head reference/chr19_genes.gtf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zSt-SqDHP3wC",
        "outputId": "898909e6-55b5-4ac9-c75a-ad33b93eaa09"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "93111 reference/chr19_genes.gtf\n",
            "chr19\tncbiRefSeq.2024-02-14\ttranscript\t61293275\t61293733\t.\t-\t.\tgene_id \"Gm20762\"; transcript_id \"Gm20762\";  gene_name \"Gm20762\";\n",
            "chr19\tncbiRefSeq.2024-02-14\texon\t61293275\t61293733\t.\t-\t.\tgene_id \"Gm20762\"; transcript_id \"Gm20762\"; exon_number \"1\"; exon_id \"Gm20762.1\"; gene_name \"Gm20762\";\n",
            "chr19\tncbiRefSeq.2024-02-14\ttranscript\t61245314\t61246460\t.\t-\t.\tgene_id \"Gm50357\"; transcript_id \"Gm50357\";  gene_name \"Gm50357\";\n",
            "chr19\tncbiRefSeq.2024-02-14\texon\t61245314\t61246460\t.\t-\t.\tgene_id \"Gm50357\"; transcript_id \"Gm50357\"; exon_number \"1\"; exon_id \"Gm50357.1\"; gene_name \"Gm50357\";\n",
            "chr19\tncbiRefSeq.2024-02-14\ttranscript\t61244246\t61247578\t.\t+\t.\tgene_id \"Gm18999\"; transcript_id \"Gm18999\";  gene_name \"Gm18999\";\n",
            "chr19\tncbiRefSeq.2024-02-14\texon\t61244246\t61247578\t.\t+\t.\tgene_id \"Gm18999\"; transcript_id \"Gm18999\"; exon_number \"1\"; exon_id \"Gm18999.1\"; gene_name \"Gm18999\";\n",
            "chr19\tncbiRefSeq.2024-02-14\ttranscript\t61220835\t61221373\t.\t-\t.\tgene_id \"Gm4242\"; transcript_id \"Gm4242\";  gene_name \"Gm4242\";\n",
            "chr19\tncbiRefSeq.2024-02-14\texon\t61220835\t61221373\t.\t-\t.\tgene_id \"Gm4242\"; transcript_id \"Gm4242\"; exon_number \"1\"; exon_id \"Gm4242.1\"; gene_name \"Gm4242\";\n",
            "chr19\tncbiRefSeq.2024-02-14\ttranscript\t61214072\t61221141\t.\t-\t.\tgene_id \"Csf2ra\"; transcript_id \"XM_036161418.1\";  gene_name \"Csf2ra\";\n",
            "chr19\tncbiRefSeq.2024-02-14\texon\t61214072\t61214090\t.\t-\t.\tgene_id \"Csf2ra\"; transcript_id \"XM_036161418.1\"; exon_number \"8\"; exon_id \"XM_036161418.1.8\"; gene_name \"Csf2ra\";\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Install STAR aligner\n",
        "!apt-get install -qq rna-star\n",
        "\n",
        "# Verify installation\n",
        "!STAR --version\n",
        "\n",
        "# Build STAR index for chr19\n",
        "# This will take ~5-10 minutes\n",
        "!STAR --runMode genomeGenerate \\\n",
        "  --genomeDir star_index \\\n",
        "  --genomeFastaFiles reference/chr19.fa \\\n",
        "  --sjdbGTFfile reference/chr19_genes.gtf \\\n",
        "  --sjdbOverhang 99 \\\n",
        "  --genomeSAindexNbases 12 \\\n",
        "  --runThreadN 2\n",
        "\n",
        "# Check what was created\n",
        "!ls -lh star_index/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "H4ilIFnVSPaE",
        "outputId": "306064a8-aad2-4a68-85aa-649f02c732d9"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Selecting previously unselected package libhtscodecs2:amd64.\n",
            "(Reading database ... \r(Reading database ... 5%\r(Reading database ... 10%\r(Reading database ... 15%\r(Reading database ... 20%\r(Reading database ... 25%\r(Reading database ... 30%\r(Reading database ... 35%\r(Reading database ... 40%\r(Reading database ... 45%\r(Reading database ... 50%\r(Reading database ... 55%\r(Reading database ... 60%\r(Reading database ... 65%\r(Reading database ... 70%\r(Reading database ... 75%\r(Reading database ... 80%\r(Reading database ... 85%\r(Reading database ... 90%\r(Reading database ... 95%\r(Reading database ... 100%\r(Reading database ... 127403 files and directories currently installed.)\n",
            "Preparing to unpack .../libhtscodecs2_1.1.1-3_amd64.deb ...\n",
            "Unpacking libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Selecting previously unselected package libhts3:amd64.\n",
            "Preparing to unpack .../libhts3_1.13+ds-2build1_amd64.deb ...\n",
            "Unpacking libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Selecting previously unselected package rna-star.\n",
            "Preparing to unpack .../rna-star_2.7.10a+dfsg-1_amd64.deb ...\n",
            "Unpacking rna-star (2.7.10a+dfsg-1) ...\n",
            "Setting up libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Setting up libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Setting up rna-star (2.7.10a+dfsg-1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero_v2.so.0 is not a symbolic link\n",
            "\n",
            "2.7.10a\n",
            "\t/usr/lib/rna-star/bin/STAR-avx2 --runMode genomeGenerate --genomeDir star_index --genomeFastaFiles reference/chr19.fa --sjdbGTFfile reference/chr19_genes.gtf --sjdbOverhang 99 --genomeSAindexNbases 12 --runThreadN 2\n",
            "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
            "Oct 30 18:02:38 ..... started STAR run\n",
            "Oct 30 18:02:38 ... starting to generate Genome files\n",
            "Oct 30 18:02:40 ..... processing annotations GTF\n",
            "!!!!! WARNING: --genomeSAindexNbases 12 is too large for the genome size=61420004, which may cause seg-fault at the mapping step. Re-run genome generation with recommended --genomeSAindexNbases 11\n",
            "Oct 30 18:02:40 ... starting to sort Suffix Array. This may take a long time...\n",
            "Oct 30 18:02:41 ... sorting Suffix Array chunks and saving them to disk...\n",
            "Oct 30 18:04:23 ... loading chunks from disk, packing SA...\n",
            "Oct 30 18:04:25 ... finished generating suffix array\n",
            "Oct 30 18:04:25 ... generating Suffix Array index\n",
            "Oct 30 18:04:43 ... completed Suffix Array index\n",
            "Oct 30 18:04:43 ..... inserting junctions into the genome indices\n",
            "Oct 30 18:05:03 ... writing Genome to disk ...\n",
            "Oct 30 18:05:04 ... writing Suffix Array to disk ...\n",
            "Oct 30 18:05:06 ... writing SAindex to disk\n",
            "Oct 30 18:05:10 ..... finished successfully\n",
            "total 629M\n",
            "-rw-r--r-- 1 root root    9 Oct 30 18:02 chrLength.txt\n",
            "-rw-r--r-- 1 root root   15 Oct 30 18:02 chrNameLength.txt\n",
            "-rw-r--r-- 1 root root    6 Oct 30 18:02 chrName.txt\n",
            "-rw-r--r-- 1 root root   11 Oct 30 18:02 chrStart.txt\n",
            "-rw-r--r-- 1 root root 1.2M Oct 30 18:02 exonGeTrInfo.tab\n",
            "-rw-r--r-- 1 root root 643K Oct 30 18:02 exonInfo.tab\n",
            "-rw-r--r-- 1 root root  43K Oct 30 18:02 geneInfo.tab\n",
            "-rw-r--r-- 1 root root  61M Oct 30 18:05 Genome\n",
            "-rw-r--r-- 1 root root  694 Oct 30 18:05 genomeParameters.txt\n",
            "-rw-r--r-- 1 root root 4.4K Oct 30 18:05 Log.out\n",
            "-rw-r--r-- 1 root root 473M Oct 30 18:05 SA\n",
            "-rw-r--r-- 1 root root  94M Oct 30 18:05 SAindex\n",
            "-rw-r--r-- 1 root root 226K Oct 30 18:04 sjdbInfo.txt\n",
            "-rw-r--r-- 1 root root 263K Oct 30 18:02 sjdbList.fromGTF.out.tab\n",
            "-rw-r--r-- 1 root root 226K Oct 30 18:04 sjdbList.out.tab\n",
            "-rw-r--r-- 1 root root 212K Oct 30 18:02 transcriptInfo.tab\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# See what files were generated\n",
        "!ls -lh star_index/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Oel_sBwTV_Fz",
        "outputId": "fbe320e8-8c77-4ba5-8811-bff11ca9cce6"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "total 629M\n",
            "-rw-r--r-- 1 root root    9 Oct 30 18:02 chrLength.txt\n",
            "-rw-r--r-- 1 root root   15 Oct 30 18:02 chrNameLength.txt\n",
            "-rw-r--r-- 1 root root    6 Oct 30 18:02 chrName.txt\n",
            "-rw-r--r-- 1 root root   11 Oct 30 18:02 chrStart.txt\n",
            "-rw-r--r-- 1 root root 1.2M Oct 30 18:02 exonGeTrInfo.tab\n",
            "-rw-r--r-- 1 root root 643K Oct 30 18:02 exonInfo.tab\n",
            "-rw-r--r-- 1 root root  43K Oct 30 18:02 geneInfo.tab\n",
            "-rw-r--r-- 1 root root  61M Oct 30 18:05 Genome\n",
            "-rw-r--r-- 1 root root  694 Oct 30 18:05 genomeParameters.txt\n",
            "-rw-r--r-- 1 root root 4.4K Oct 30 18:05 Log.out\n",
            "-rw-r--r-- 1 root root 473M Oct 30 18:05 SA\n",
            "-rw-r--r-- 1 root root  94M Oct 30 18:05 SAindex\n",
            "-rw-r--r-- 1 root root 226K Oct 30 18:04 sjdbInfo.txt\n",
            "-rw-r--r-- 1 root root 263K Oct 30 18:02 sjdbList.fromGTF.out.tab\n",
            "-rw-r--r-- 1 root root 226K Oct 30 18:04 sjdbList.out.tab\n",
            "-rw-r--r-- 1 root root 212K Oct 30 18:02 transcriptInfo.tab\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Run STAR alignment on our trimmed test sample\n",
        "!STAR --genomeDir star_index \\\n",
        "  --readFilesIn fastq_files/SRR21820609_1_trimmed.fastq \\\n",
        "                fastq_files/SRR21820609_2_trimmed.fastq \\\n",
        "  --outFileNamePrefix alignment_output/sample_ \\\n",
        "  --outSAMtype BAM SortedByCoordinate \\\n",
        "  --runThreadN 2 \\\n",
        "  --quantMode GeneCounts\n",
        "\n",
        "# Check what was created\n",
        "!ls -lh alignment_output/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "e9J_43V9WejQ",
        "outputId": "a1383513-943f-4d24-bc46-c290988d33a6"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\t/usr/lib/rna-star/bin/STAR-avx2 --genomeDir star_index --readFilesIn fastq_files/SRR21820609_1_trimmed.fastq fastq_files/SRR21820609_2_trimmed.fastq --outFileNamePrefix alignment_output/sample_ --outSAMtype BAM SortedByCoordinate --runThreadN 2 --quantMode GeneCounts\n",
            "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
            "Oct 30 18:20:56 ..... started STAR run\n",
            "Oct 30 18:20:56 ..... loading genome\n",
            "Oct 30 18:20:59 ..... started mapping\n",
            "Oct 30 18:44:13 ..... finished mapping\n",
            "Oct 30 18:44:13 ..... started sorting BAM\n",
            "Oct 30 18:44:13 ..... finished successfully\n",
            "total 40K\n",
            "-rw-r--r-- 1 root root  652 Oct 30 18:44 sample_Aligned.sortedByCoord.out.bam\n",
            "-rw-r--r-- 1 root root 2.0K Oct 30 18:44 sample_Log.final.out\n",
            "-rw-r--r-- 1 root root 5.0K Oct 30 18:44 sample_Log.out\n",
            "-rw-r--r-- 1 root root  482 Oct 30 18:44 sample_Log.progress.out\n",
            "-rw-r--r-- 1 root root  19K Oct 30 18:44 sample_ReadsPerGene.out.tab\n",
            "-rw-r--r-- 1 root root    0 Oct 30 18:44 sample_SJ.out.tab\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Check alignment quality\n",
        "!cat alignment_output/sample_Log.final.out"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MOuHIT3Ke7MB",
        "outputId": "2f934ef0-508d-4065-8e26-e22ce2bc334b"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "                                 Started job on |\tOct 30 18:20:56\n",
            "                             Started mapping on |\tOct 30 18:20:59\n",
            "                                    Finished on |\tOct 30 18:44:13\n",
            "       Mapping speed, Million of reads per hour |\t0.25\n",
            "\n",
            "                          Number of input reads |\t95919\n",
            "                      Average input read length |\t291\n",
            "                                    UNIQUE READS:\n",
            "                   Uniquely mapped reads number |\t1\n",
            "                        Uniquely mapped reads % |\t0.00%\n",
            "                          Average mapped length |\t252.00\n",
            "                       Number of splices: Total |\t2\n",
            "            Number of splices: Annotated (sjdb) |\t0\n",
            "                       Number of splices: GT/AG |\t2\n",
            "                       Number of splices: GC/AG |\t0\n",
            "                       Number of splices: AT/AC |\t0\n",
            "               Number of splices: Non-canonical |\t0\n",
            "                      Mismatch rate per base, % |\t3.97%\n",
            "                         Deletion rate per base |\t0.00%\n",
            "                        Deletion average length |\t0.00\n",
            "                        Insertion rate per base |\t0.00%\n",
            "                       Insertion average length |\t0.00\n",
            "                             MULTI-MAPPING READS:\n",
            "        Number of reads mapped to multiple loci |\t0\n",
            "             % of reads mapped to multiple loci |\t0.00%\n",
            "        Number of reads mapped to too many loci |\t0\n",
            "             % of reads mapped to too many loci |\t0.00%\n",
            "                                  UNMAPPED READS:\n",
            "  Number of reads unmapped: too many mismatches |\t0\n",
            "       % of reads unmapped: too many mismatches |\t0.00%\n",
            "            Number of reads unmapped: too short |\t95890\n",
            "                 % of reads unmapped: too short |\t99.97%\n",
            "                Number of reads unmapped: other |\t28\n",
            "                     % of reads unmapped: other |\t0.03%\n",
            "                                  CHIMERIC READS:\n",
            "                       Number of chimeric reads |\t0\n",
            "                            % of chimeric reads |\t0.00%\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "tECqWs3PiWyo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# First 20 lines of gene counts\n",
        "!head -20 alignment_output/sample_ReadsPerGene.out.tab"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OMtiBm8hfCo6",
        "outputId": "0cf064b4-fbe2-44e9-f92c-cd7a57589bbc"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "N_unmapped\t95918\t95918\t95918\n",
            "N_multimapping\t0\t0\t0\n",
            "N_noFeature\t1\t1\t1\n",
            "N_ambiguous\t0\t0\t0\n",
            "Gm20762\t0\t0\t0\n",
            "Gm50357\t0\t0\t0\n",
            "Gm18999\t0\t0\t0\n",
            "Gm4242\t0\t0\t0\n",
            "Csf2ra\t0\t0\t0\n",
            "Mplkipl1\t0\t0\t0\n",
            "Gm5521\t0\t0\t0\n",
            "Gm22365\t0\t0\t0\n",
            "Gm6020\t0\t0\t0\n",
            "Gm4238\t0\t0\t0\n",
            "Zfp950\t0\t0\t0\n",
            "Gm22520\t0\t0\t0\n",
            "Gm9529\t0\t0\t0\n",
            "Grk5\t0\t0\t0\n",
            "Prdx3\t0\t0\t0\n",
            "Sfxn4\t0\t0\t0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Look at what DID map\n",
        "!samtools view alignment_output/sample_Aligned.sortedByCoord.out.bam | head -5\n",
        "\n",
        "# And look at the gene counts\n",
        "!head -30 alignment_output/sample_ReadsPerGene.out.tab"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-0hl8yEkiZM5",
        "outputId": "79a7214a-1c99-49e3-bda2-ff509b066e11"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/bin/bash: line 1: samtools: command not found\n",
            "N_unmapped\t95918\t95918\t95918\n",
            "N_multimapping\t0\t0\t0\n",
            "N_noFeature\t1\t1\t1\n",
            "N_ambiguous\t0\t0\t0\n",
            "Gm20762\t0\t0\t0\n",
            "Gm50357\t0\t0\t0\n",
            "Gm18999\t0\t0\t0\n",
            "Gm4242\t0\t0\t0\n",
            "Csf2ra\t0\t0\t0\n",
            "Mplkipl1\t0\t0\t0\n",
            "Gm5521\t0\t0\t0\n",
            "Gm22365\t0\t0\t0\n",
            "Gm6020\t0\t0\t0\n",
            "Gm4238\t0\t0\t0\n",
            "Zfp950\t0\t0\t0\n",
            "Gm22520\t0\t0\t0\n",
            "Gm9529\t0\t0\t0\n",
            "Grk5\t0\t0\t0\n",
            "Prdx3\t0\t0\t0\n",
            "Sfxn4\t0\t0\t0\n",
            "LOC118568411\t0\t0\t0\n",
            "Dennd10\t0\t0\t0\n",
            "Snora19\t0\t0\t0\n",
            "Mir3084-2\t0\t0\t0\n",
            "Eif3a\t0\t0\t0\n",
            "Nanos1\t0\t0\t0\n",
            "Gm34406\t0\t0\t0\n",
            "Gm34459\t0\t0\t0\n",
            "Gm46652\t0\t0\t0\n",
            "Gm41884\t0\t0\t0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Clean up and start fresh\n",
        "!rm -rf star_index/*\n",
        "!rm -rf reference/chr19*\n",
        "\n",
        "# Download chromosome X\n",
        "!wget -q -O reference/chrX.fa.gz \\\n",
        "  \"http://hgdownload.soe.ucsc.edu/goldenPath/mm39/chromosomes/chrX.fa.gz\"\n",
        "\n",
        "# Uncompress\n",
        "!gunzip reference/chrX.fa.gz\n",
        "\n",
        "# Filter GTF to chrX genes\n",
        "!grep \"^chrX\" reference/mm39_genes.gtf > reference/chrX_genes.gtf\n",
        "\n",
        "# Check what we got\n",
        "!ls -lh reference/\n",
        "!wc -l reference/chrX_genes.gtf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hwcslhXNkRCF",
        "outputId": "5592420d-43fd-46c6-9574-2720dcbd0061"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "total 739M\n",
            "-rw-r--r-- 1 root root 165M Sep 10  2020 chrX.fa\n",
            "-rw-r--r-- 1 root root  19M Oct 30 19:21 chrX_genes.gtf\n",
            "-rw-r--r-- 1 root root 556M Feb 14  2024 mm39_genes.gtf\n",
            "116446 reference/chrX_genes.gtf\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Build STAR index for chrX\n",
        "!STAR --runMode genomeGenerate \\\n",
        "  --genomeDir star_index \\\n",
        "  --genomeFastaFiles reference/chrX.fa \\\n",
        "  --sjdbGTFfile reference/chrX_genes.gtf \\\n",
        "  --sjdbOverhang 99 \\\n",
        "  --genomeSAindexNbases 13 \\\n",
        "  --runThreadN 2"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "L7fOx9y5k0iy",
        "outputId": "24c8c156-0a33-449a-c760-3971f6007659"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\t/usr/lib/rna-star/bin/STAR-avx2 --runMode genomeGenerate --genomeDir star_index --genomeFastaFiles reference/chrX.fa --sjdbGTFfile reference/chrX_genes.gtf --sjdbOverhang 99 --genomeSAindexNbases 13 --runThreadN 2\n",
            "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
            "Oct 30 19:23:45 ..... started STAR run\n",
            "Oct 30 19:23:45 ... starting to generate Genome files\n",
            "Oct 30 19:23:49 ..... processing annotations GTF\n",
            "!!!!! WARNING: --genomeSAindexNbases 13 is too large for the genome size=169476592, which may cause seg-fault at the mapping step. Re-run genome generation with recommended --genomeSAindexNbases 12\n",
            "Oct 30 19:23:50 ... starting to sort Suffix Array. This may take a long time...\n",
            "Oct 30 19:23:51 ... sorting Suffix Array chunks and saving them to disk...\n",
            "Oct 30 19:30:33 ... loading chunks from disk, packing SA...\n",
            "Oct 30 19:30:46 ... finished generating suffix array\n",
            "Oct 30 19:30:46 ... generating Suffix Array index\n",
            "Oct 30 19:31:28 ... completed Suffix Array index\n",
            "Oct 30 19:31:28 ..... inserting junctions into the genome indices\n",
            "Oct 30 19:31:56 ... writing Genome to disk ...\n",
            "Oct 30 19:31:56 ... writing Suffix Array to disk ...\n",
            "Oct 30 19:32:12 ... writing SAindex to disk\n",
            "Oct 30 19:32:15 ..... finished successfully\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Align to chrX\n",
        "!STAR --genomeDir star_index \\\n",
        "  --readFilesIn fastq_files/SRR21820609_1_trimmed.fastq \\\n",
        "                fastq_files/SRR21820609_2_trimmed.fastq \\\n",
        "  --outFileNamePrefix alignment_output/sampleX_ \\\n",
        "  --outSAMtype BAM SortedByCoordinate \\\n",
        "  --runThreadN 2 \\\n",
        "  --quantMode GeneCounts\n",
        "\n",
        "# Check results\n",
        "!cat alignment_output/sampleX_Log.final.out | grep \"Uniquely mapped\"\n",
        "!head -20 alignment_output/sampleX_ReadsPerGene.out.tab"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "q08cqyrQlBLl",
        "outputId": "a97c753e-7b7c-46f0-b010-efaede8b4c74"
      },
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\t/usr/lib/rna-star/bin/STAR-avx2 --genomeDir star_index --readFilesIn fastq_files/SRR21820609_1_trimmed.fastq fastq_files/SRR21820609_2_trimmed.fastq --outFileNamePrefix alignment_output/sampleX_ --outSAMtype BAM SortedByCoordinate --runThreadN 2 --quantMode GeneCounts\n",
            "\tSTAR version: 2.7.10a   compiled: 2022-01-16T16:35:44+00:00 <place not set in Debian package>\n",
            "Oct 30 19:36:53 ..... started STAR run\n",
            "Oct 30 19:36:53 ..... loading genome\n",
            "Oct 30 19:36:55 ..... started mapping\n",
            "Oct 30 19:56:09 ..... finished mapping\n",
            "Oct 30 19:56:09 ..... started sorting BAM\n",
            "Oct 30 19:56:09 ..... finished successfully\n",
            "                   Uniquely mapped reads number |\t0\n",
            "                        Uniquely mapped reads % |\t0.00%\n",
            "N_unmapped\t95919\t95919\t95919\n",
            "N_multimapping\t0\t0\t0\n",
            "N_noFeature\t0\t0\t0\n",
            "N_ambiguous\t0\t0\t0\n",
            "Asmt\t0\t0\t0\n",
            "Akap17a\t0\t0\t0\n",
            "2510022D24Rik\t0\t0\t0\n",
            "Nlgn4l\t0\t0\t0\n",
            "Sts\t0\t0\t0\n",
            "Erdr1x\t0\t0\t0\n",
            "Gm15247\t0\t0\t0\n",
            "G530011O06Rik\t0\t0\t0\n",
            "Gm54065\t0\t0\t0\n",
            "Capza1b\t0\t0\t0\n",
            "Gm25651\t0\t0\t0\n",
            "Mid1\t0\t0\t0\n",
            "Gm15245\t0\t0\t0\n",
            "Gm15246\t0\t0\t0\n",
            "Hccs\t0\t0\t0\n",
            "Amelx\t0\t0\t0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Check a few raw reads to see what they look like\n",
        "!head -8 fastq_files/SRR21820609_1_trimmed.fastq\n",
        "\n",
        "# And check if these SRR numbers are actually from your project\n",
        "!echo \"SRR21820609\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "YR8FQdFe0_Pc",
        "outputId": "d091d388-8513-4efa-f7b5-f48beca6ca84"
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "@SRR21820609.3 3 length=151\n",
            "NCTTCAGTCAGCTTCCGCGCCAGCAAGATACGGACGGCGAAGCCAAAGCAATACGAAACAAGCGAGAAAAGCCCTACATATTCTATTACTCAAGGCTNAACTTATTGAAAACTAAAGCGCTAACGAGCAAGAAAACGGGAAGGTTGGTTCG\n",
            "+SRR21820609.3 3 length=151\n",
            "#AAFFJJFJJJJJJAJJJFJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJ-AJJJJJJJJJJJJJJJJJJJJJA<J--AFJJJJJJJJJAFJF-<-AJ#JFJJJJJFJ-FJJJ<FJAJFF-FAJJJ7<-FF7FFAJFAFJAF-FFJFAJF7-\n",
            "@SRR21820609.4 4 length=151\n",
            "NCTTCAGTCTGCTGGTTATATAGAGTGATTGAAACCAGTTAGAGCCGGGGGGAGTAATTCTCCTCCTTCCCCACCTTTCGCCCTTTCCACACCCAACNCCCTCCATTACATAAAGCATTGTTGACATCCATCGGCTACATAGGCTATTTTA\n",
            "+SRR21820609.4 4 length=151\n",
            "#AAAFJJJJJJJJJJJJFJJJFJJJFJJJFFJJFFFAJJFJFJFFJAJJ<-A<J<A<<JFAAJJJJJJJJJJJ-AAAJ-<JJJJJJJJJJJJJJ<J-#FJJJJ-<JF<<AJJJJFFF7<-<--7FJ7A-7--7A-A-FAA-<FJJAFFAF-\n",
            "SRR21820609\n"
          ]
        }
      ]
    }
  ]
}